### PR TITLE
Login

### PR DESCRIPTION
**Pull request by Nosoyleon**
## Summary
1. Adding Log in Form with redux-form
2. Vaidating form
3. Using react-router-dom for navigation

## Screenshots
image

## Trello Card
https://trello.com/c/w6BurH58/9-routing-i-login

### CODE REVIEW

File: react/training-project/src/app/components/Input/index.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types';
+
+import style from './styles.scss';
+
+function Input({ placeholder, input, type, meta }) {
+  return (
+    <Fragment>
+      <input className={style.input} {...input} placeholder={placeholder} type={type} />
+      {meta.error && meta.touched && <div className={style.fieldError}>{meta.error}</div>}
```
Comment (BJordi) [inline react/training-project/src/app/components/Input/index.js]: Don't use divs to show text. Instead, use text tags like span or h1-h6.

File: react/training-project/src/app/components/Input/styles.scss [hunk 2]
```diff
@@ -0,0 +1,12 @@
+@import 'src/scss/fixed-sizes';
+@import 'src/scss/colors';
+
+.field-error {
+  color: $font-danger;
+}
+.input {
```
Comment (BJordi) [inline react/training-project/src/app/components/Input/styles.scss:8]: Add an empty line above this one.

File: react/training-project/src/app/screens/Game/layout.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import PropTypes from 'prop-types';
+
+import Board from './components/Board';
+import style from './styles.scss';
+
+function GameDumb({ current, winner, xIsNext, moves }) {
+  return (
+    <div className={style.game}>
+      <div className={style.gameBoard}>
+        <Board squares={current.squares} />
```
Comment (BJordi) [inline react/training-project/src/app/screens/Game/layout.js:11]: If current is not required, than, it's possible that current has a null or undefined value, right? If that's the case, `current.squares` may fail.

File: react/training-project/src/app/screens/Login/index.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { Redirect } from 'react-router-dom';
+import PropTypes from 'prop-types';
+
+import actionCreators from '~redux/Auth/actions';
+
+import LoginForm from './layout';
+
+/* eslint-disable no-func-assign */
+class LoginPage extends Component {
+  submit = vals => {
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/index.js]: values?

Comment (BJordi) [inline react/training-project/src/app/screens/Login/index.js]: We usually call this functions in this way `handleMyFunction`. Example: If the function will be used when clicking, then you can call it `handleClick` and then do something like this `onClick={this.handleClick}`.

File: react/training-project/src/app/screens/Login/styles.scss [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+.login-button {
+  background-color: $primary;
+  color: $white;
+  cursor: pointer;
+  font-size: $large;
+  font-weight: bold;
+  margin-top: 30px;
+  padding: 10px;
+  width: 100%;
+
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/styles.scss]: Remove this line.

File: react/training-project/src/redux/Auth/actions.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  LOG_IN_PROCESSING: '@AUTH/LOG_IN_PROCESSING',
+  LOG_OUT: '@@AUTH/V',
+  LOG_OUT_SUCCESS: '@@AUTH/LOG_OUT_SUCCESS',
+  LOG_OUT_FAIL: '@@AUTH/LOG_OUT_FAIL'
+};
+
+const actionCreators = {
+  logIn: user => dispatch => {
+    dispatch(actionCreators.logInProcessing(true));
+    setTimeout(async () => {
```
Comment (BJordi) [inline react/training-project/src/redux/Auth/actions.js]: Why are you using this setTimeout here? :thinking:

File: react/training-project/src/redux/Auth/actions.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      const token = await AuthService.logIn(user);
+      if (!token) return dispatch(actionCreators.logInFail(true, false));
+      dispatch(actionCreators.logInSuccess(token, false));
+    }, 1500);
+  },
+
+  logInProcessing: bool => ({
```
Comment (BJordi) [inline react/training-project/src/redux/Auth/actions.js:14]: Just a NTH, you can extract this function and the other two below in another object called `privateActionCreators`.

File: react/training-project/src/services/AuthService.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+      const response = await api.post('/sessions', { token });
+      answer = response.ok ? response.data.token : null;
+    }
+
+    return answer;
+  },
+
+  getUserToken: async token => {
+    const response = await api.get('/sessions', { token });
+    return response.ok ? response.data.token : null;
```
Comment (BJordi) [inline react/training-project/src/services/AuthService.js]: What we usually do is to just make the call to the API and in the action deal with the response.
```
//Service file
getUserToken: () => api.get('/sessions', { token });

//Actions file
const response = AuthService.getUserToken(token);
if (response.ok) {
  // Do what you have to do with the response.data
} else {
  // Do what you have to do if the response status is different to 2XX
}
```

**Review by BJordi — CHANGES_REQUESTED** [review]

File: react/training-project/src/redux/Auth/actions.js (hunk 6 above)
Comment (Nosoyleon) [reply react/training-project/src/redux/Auth/actions.js]: Just emulating a delay in API response 😄

File: react/training-project/src/app/screens/Game/index.js [hunk 9]
```diff
@@ -27,33 +27,28 @@ class Game extends Component {
...
-          <h3 className={style.status}>
-            {winner ? `Winner: ${winner}` : `Next Player ${this.props.xIsNext ? 'X' : 'O'}`}
-          </h3>
-          <ol>{moves}</ol>
-        </div>
-      </div>
+    return this.props.logged ? (
+      <GameDumb current={current} winner={winner} xIsNext={this.props.xIsNext} moves={moves} />
+    ) : (
+      <Redirect to="/login" />
```
Comment (BJordi) [inline react/training-project/src/app/screens/Game/index.js:33]: Have a constants file in the routes folder and add this one there. Do it in another PR.

File: react/training-project/src/app/screens/Game/layout.js (hunk 3 above)
Comment (BJordi) [reply react/training-project/src/app/screens/Game/layout.js:11]: ...? :thinking: :eyes:

File: react/training-project/src/app/screens/Login/index.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+  render() {
+    return !this.props.token ? (
+      <LoginForm
+        processing={this.props.processing}
+        loginFail={this.props.loginFail}
+        onSubmit={this.handleSubmit}
+      />
+    ) : (
+      <Redirect to="/" />
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/index.js:24]: Same I mentioned above.

File: react/training-project/src/app/screens/Login/layout.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+LoginForm.propTypes = {
+  handleSubmit: PropTypes.func.isRequired,
+  loginFail: PropTypes.bool.isRequired,
+  processing: PropTypes.bool.isRequired
+};
+
+LoginForm = reduxForm({
+  form: 'login'
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/layout.js:49]: For future projects, if you are going to use many forms, you can add it to a formNames file :wink::ok_hand:

**Review by BJordi — APPROVED** [review]

**Pull request merged**